rinja/src/filters/json.rs

```diff
@@ -239,7 +240,10 @@ impl<W: fmt::Write + ?Sized> io::Write for JsonWriter<'_, W> {
 
     #[inline]
     fn write_all(&mut self, bytes: &[u8]) -> io::Result<()> {
-        write(self.0, bytes).map_err(|err| io::Error::new(io::ErrorKind::InvalidData, err))
+        // SAFETY: `serde_json` only writes valid strings
```

Comment:
This invariant is still rather fragile because it dpeends on how `JsonWriter` is used, and verifying that is not a super local task.

It seems unfortunate that `serde_json` supports `io::Write` not `fmt::Write`, but what I would do is produce a function that behaves like `Serializer::with_formatter` but for `fmt::Write` and performs the unchecked conversion inside it, so it's properly encapsulated.

Reply:
The documentation of [`serde_json::to_writer`](https://docs.rs/serde_json/1.0.138/serde_json/fn.to_writer.html) states:

> Serialization guarantees it only feeds valid UTF-8 sequences to the write.

I guess the library [was audited](https://lib.rs/crates/serde_json/audit) by enough people we can depend on this claim being true. What do you say?

But yes, it would have liked it better, too, if `serde_json` accepted `fmt::Write`, not only for safety reasons, but also because `io::Write` cannot be used in `#![no_std]`.

Reply:
Eh, the audits won't have checked that type of invariant, but it's okay enough to rely on here IMO. I'm mostly suggesting that it be encapsulated so that the assumption of only passing serde-json writers is local rather than dependent on the caller of this code.